File: packages/inputlayer-py/src/inputlayer/notifications.py
"""Notification dispatcher for push events from the server."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Callable


@dataclass(frozen=True)
class NotificationEvent:
    """A single notification event from the server."""

    type: str  # persistent_update, rule_change, kg_change, schema_change
    seq: int
    timestamp_ms: int
    session_id: str | None = None
    knowledge_graph: str | None = None
    relation: str | None = None
    operation: str | None = None
    count: int | None = None
    rule_name: str | None = None
    entity: str | None = None


Callback = Callable[[NotificationEvent], Any]


class NotificationDispatcher:
    """Routes notification events to registered callbacks."""

    def __init__(self) -> None:
        self._callbacks: list[tuple[str | None, str | None, str | None, Callback]] = []
        self._queue: asyncio.Queue[NotificationEvent] = asyncio.Queue()
        self._last_seq: int = 0

    @property
    def last_seq(self) -> int:
        return self._last_seq

    def on(
        self,
        event_type: str | None = None,
        *,
        relation: str | None = None,
        knowledge_graph: str | None = None,
        callback: Callback | None = None,
    ) -> Callable | None:
        """Register a callback for notifications. Can be used as a decorator."""
        def decorator(fn: Callback) -> Callback:
            self._callbacks.append((event_type, relation, knowledge_graph, fn))
            return fn

        if callback is not None:
            self._callbacks.append((event_type, relation, knowledge_graph, callback))
            return None
        return decorator

    def dispatch(self, event: NotificationEvent) -> None:
        """Dispatch a notification to matching callbacks and the async queue."""
        self._last_seq = max(self._last_seq, event.seq)
        # Push to async iterator queue
        self._queue.put_nowait(event)
        # Call matching callbacks
        for evt_type, rel, kg, cb in self._callbacks:
            if evt_type is not None and event.type != evt_type:
                continue
            if rel is not None and event.relation != rel:
                continue
            if kg is not None and event.knowledge_graph != kg:
                continue
            try:
                result = cb(event)
                if asyncio.iscoroutine(result):
                    asyncio.ensure_future(result)
            except Exception:
                pass  # Callbacks should not break the dispatcher

    async def __aiter__(self) -> AsyncIterator[NotificationEvent]:
        """Async iterator yielding notification events."""
        while True:
            event = await self._queue.get()
            yield event

File: packages/inputlayer-js/src/notifications.ts
/**
 * Notification dispatcher for push events from the server.
 */

/** A single notification event from the server. */
export interface NotificationEvent {
  type: string;
  seq: number;
  timestampMs: number;
  sessionId?: string;
  knowledgeGraph?: string;
  relation?: string;
  operation?: string;
  count?: number;
  ruleName?: string;
  entity?: string;
}

export type NotificationCallback = (event: NotificationEvent) => void | Promise<void>;

interface CallbackEntry {
  eventType?: string;
  relation?: string;
  knowledgeGraph?: string;
  callback: NotificationCallback;
}

/**
 * Routes notification events to registered callbacks.
 */
export class NotificationDispatcher {
  private callbacks: CallbackEntry[] = [];
  private _lastSeq = 0;
  private waiters: Array<(event: NotificationEvent) => void> = [];

  get lastSeq(): number {
    return this._lastSeq;
  }

  /**
   * Register a callback for notifications.
   *
   * @param eventType - Filter by event type (e.g. "persistent_update")
   * @param opts - Additional filters
   * @param callback - Function to call when a matching event arrives
   */
  on(
    eventType: string | undefined,
    opts: { relation?: string; knowledgeGraph?: string },
    callback: NotificationCallback,
  ): void {
    this.callbacks.push({
      eventType,
      relation: opts.relation,
      knowledgeGraph: opts.knowledgeGraph,
      callback,
    });
  }

  /** Remove a previously registered callback. */
  off(callback: NotificationCallback): void {
    this.callbacks = this.callbacks.filter((e) => e.callback !== callback);
  }

  /** Dispatch a notification to matching callbacks. */
  dispatch(event: NotificationEvent): void {
    this._lastSeq = Math.max(this._lastSeq, event.seq);

    // Notify any async iterators
    for (const waiter of this.waiters) {
      waiter(event);
    }
    this.waiters = [];

    // Call matching callbacks
    for (const entry of this.callbacks) {
      if (entry.eventType !== undefined && event.type !== entry.eventType) continue;
      if (entry.relation !== undefined && event.relation !== entry.relation) continue;
      if (entry.knowledgeGraph !== undefined && event.knowledgeGraph !== entry.knowledgeGraph) continue;
      try {
        entry.callback(event);
      } catch {
        // Callbacks should not break the dispatcher
      }
    }
  }

  /** Wait for the next notification event. */
  next(): Promise<NotificationEvent> {
    return new Promise<NotificationEvent>((resolve) => {
      this.waiters.push(resolve);
    });
  }

  /**
   * Create an async iterable of notification events.
   *
   * Usage:
   *   for await (const event of dispatcher.events()) { ... }
   */
  async *events(): AsyncIterableIterator<NotificationEvent> {
    while (true) {
      yield await this.next();
    }
  }
}
